File: claude-code/scrapy_researchers/bloom_filter.py
"""Bloom filter for Scrapy URL deduplication using rbloom.

rbloom is a Rust-backed bloom filter — the fastest Python implementation
(2x faster than pybloomfiltermmap3, 10x faster than pybloom-live).

No Redis dependency — designed for single-node iterative crawling where
the full SHA256 fingerprint set grows too large across rounds.

Key properties:
- No false negatives: if it says "not seen", it truly hasn't been seen
- Rare false positives: configurable via expected_items and fp_rate
- Persistent: save/load to disk for cross-run deduplication
- Scrapy integration: BloomDupeFilter replaces RFPDupeFilter with optional Neon Postgres persistence

Spider-specific profiles:
- platform/claude_com: conservative FP rate, smaller capacity (doc sites)
- anthropic: large capacity (sitemap discovers many URLs)
- github: item-level only (API-driven, no HTTP dedup needed)
- llms_full: minimal (single-file parse, no link following)
- docs: moderate defaults (generic doc sites)
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass
from typing import Any

from rbloom import Bloom
from scrapy.dupefilters import BaseDupeFilter
from scrapy.http import Request
from scrapy.utils.request import fingerprint


def _stable_hash(item: object) -> int:
    """Deterministic hash function for rbloom serialization.

    rbloom requires a custom hash function (not Python's built-in hash())
    to support save_bytes/load_bytes persistence across processes.
    """
    return int.from_bytes(hashlib.sha256(str(item).encode()).digest()[:8], "big")


# ── Scrapy DupeFilter integration ─────────────────────────────────


class BloomDupeFilter(BaseDupeFilter):
    """Scrapy DupeFilter backed by rbloom with optional Neon Postgres persistence.

    Settings:
        BLOOM_EXPECTED_ITEMS (int): Expected number of unique URLs. Default: 10000.
        BLOOM_FALSE_POSITIVE_RATE (float): Target FP rate. Default: 0.001.
        BLOOM_DB_URL (str | None): Neon Postgres connection string for persistence.
        BLOOM_CRAWLER_ID (str): Identifier for this crawler instance. Default: 'scrapy-docs'.
        BLOOM_DOMAIN (str): Domain being crawled. Default: 'docs.anthropic.com'.
    """

    def __init__(
        self,
        expected_items: int = 10000,
        false_positive_rate: float = 0.001,
        db_url: str | None = None,
        crawler_id: str = "scrapy-docs",
        domain: str = "docs.anthropic.com",
    ) -> None:
        self.bloom = Bloom(expected_items, false_positive_rate, _stable_hash)
        self.db_url = db_url
        self.crawler_id = crawler_id
        self.domain = domain
        self._expected_items = expected_items
        self._false_positive_rate = false_positive_rate

    @classmethod
    def from_settings(cls, settings: Any) -> BloomDupeFilter:
        return cls(
            expected_items=settings.getint("BLOOM_EXPECTED_ITEMS", 10000),
            false_positive_rate=settings.getfloat("BLOOM_FALSE_POSITIVE_RATE", 0.001),
            db_url=settings.get("BLOOM_DB_URL"),
            crawler_id=settings.get("BLOOM_CRAWLER_ID", "scrapy-docs"),
            domain=settings.get("BLOOM_DOMAIN", "docs.anthropic.com"),
        )

    def request_seen(self, request: Request) -> bool:
        """Return True if request URL has been seen (bloom filter check)."""
        fp = fingerprint(request).hex()
        if fp in self.bloom:
            return True
        self.bloom.add(fp)
        return False

    def open(self) -> None:
        """Load bloom filter state from Neon Postgres on spider open."""
        if self.db_url:
            self._load_from_db()

    def close(self, reason: str = "") -> None:
        """Save bloom filter state to Neon Postgres on spider close."""
        if self.db_url:
            self._save_to_db()

    def _load_from_db(self) -> None:
        """Load persisted bloom filter from Neon Postgres."""
        try:
            import psycopg2  # type: ignore[import-untyped]

            conn = psycopg2.connect(self.db_url)
            cur = conn.cursor()
            cur.execute(
                "SELECT filter_bytes FROM bloom_filter_state WHERE crawler_id = %s AND domain = %s",
                (self.crawler_id, self.domain),
            )
            row = cur.fetchone()
            if row:
                self.bloom = Bloom.load_bytes(bytes(row[0]), _stable_hash)
            cur.close()
            conn.close()
        except Exception:
            pass  # Fall back to empty filter

    def _save_to_db(self) -> None:
        """Persist bloom filter to Neon Postgres."""
        try:
            import psycopg2  # type: ignore[import-untyped]

            filter_data = self.bloom.save_bytes()
            conn = psycopg2.connect(self.db_url)
            cur = conn.cursor()
            cur.execute(
                """INSERT INTO bloom_filter_state
                     (crawler_id, domain, expected_items, false_positive_rate,
                      hash_functions, bit_array_size, items_inserted, filter_bytes)
                   VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
                   ON CONFLICT (crawler_id, domain) DO UPDATE SET
                     items_inserted = EXCLUDED.items_inserted,
                     filter_bytes = EXCLUDED.filter_bytes,
                     updated_at = NOW()""",
                (
                    self.crawler_id,
                    self.domain,
                    self._expected_items,
                    self._false_positive_rate,
                    0,  # rbloom manages hash count internally
                    0,  # rbloom manages bit size internally
                    0,
                    filter_data,
                ),
            )
            conn.commit()
            cur.close()
            conn.close()
        except Exception:
            pass


# ── Spider-specific profiles ────────────────────────────────────


@dataclass(frozen=True)
class BloomProfile:
    """Spider-specific bloom filter configuration.

    Each profile tunes expected_urls, fp_rate, and persist paths based on
    the crawl pattern of the spider:
    - High-volume sitemaps need large capacity, tolerate higher FP
    - Doc sites need strict FP to avoid missing updated pages
    - Single-file parsers need minimal bloom overhead
    """

    expected_urls: int
    fp_rate: float
    persist_dir: str
    request_dedup: bool = True   # whether to use BloomDupeFilter for requests
    item_dedup: bool = True      # whether to use bloom-backed DedupPipeline
    download_delay: float = 2.0
    autothrottle_start: float = 2.0
    autothrottle_max: float = 60.0
    retry_times: int = 5

    def dupefilter_path(self, spider_name: str) -> str:
        return f"{self.persist_dir}/{spider_name}_dupefilter.bloom"

    def dedup_path(self, spider_name: str) -> str:
        return f"{self.persist_dir}/{spider_name}_dedup.bloom"

    def to_scrapy_settings(self, spider_name: str) -> dict[str, Any]:
        """Convert profile to Scrapy settings dict for spider custom_settings."""
        settings: dict[str, Any] = {
            "BLOOM_EXPECTED_URLS": self.expected_urls,
            "BLOOM_FP_RATE": self.fp_rate,
            "BLOOM_PERSIST_PATH": self.dupefilter_path(spider_name),
            "BLOOM_DEDUP_EXPECTED_ITEMS": self.expected_urls,
            "BLOOM_DEDUP_FP_RATE": self.fp_rate,
            "BLOOM_DEDUP_PERSIST_PATH": self.dedup_path(spider_name),
            "DOWNLOAD_DELAY": self.download_delay,
            "AUTOTHROTTLE_START_DELAY": self.autothrottle_start,
            "AUTOTHROTTLE_MAX_DELAY": self.autothrottle_max,
            "RETRY_TIMES": self.retry_times,
        }
        if not self.request_dedup:
            settings["DUPEFILTER_CLASS"] = "scrapy.dupefilters.RFPDupeFilter"
        return settings


# ── Spider-specific profiles ────────────────────────────────────
# Tuned based on each spider's crawl pattern and target site behavior.

BLOOM_PROFILES: dict[str, BloomProfile] = {
    # platform.claude.com: strict dedup, conservative politeness to avoid blocks.
    "platform_spider": BloomProfile(
        expected_urls=5_000,
        fp_rate=0.0005,
        persist_dir="scrapy_researchers/.bloomstate",
        download_delay=3.0,
        autothrottle_start=3.0,
        autothrottle_max=120.0,
        retry_times=7,
    ),

    # code.claude.com: moderate. Generic docs spider, ~100 pages per llms.txt.
    "docs_spider": BloomProfile(
        expected_urls=10_000,
        fp_rate=0.001,
        persist_dir="scrapy_researchers/.bloomstate",
        download_delay=2.0,
        autothrottle_start=2.0,
        autothrottle_max=60.0,
    ),

    # anthropic.com sitemap: high volume.
    "anthropic_spider": BloomProfile(
        expected_urls=50_000,
        fp_rate=0.002,
        persist_dir="scrapy_researchers/.bloomstate",
        download_delay=2.0,
        autothrottle_start=2.0,
        autothrottle_max=60.0,
    ),

    # claude.com: similar to platform — doc site, avoid blocks.
    "claude_com_spider": BloomProfile(
        expected_urls=5_000,
        fp_rate=0.0005,
        persist_dir="scrapy_researchers/.bloomstate",
        download_delay=3.0,
        autothrottle_start=3.0,
        autothrottle_max=120.0,
        retry_times=7,
    ),

    # llms-full.txt: single file download, no link following.
    "llms_full_spider": BloomProfile(
        expected_urls=1_000,
        fp_rate=0.001,
        persist_dir="scrapy_researchers/.bloomstate",
        request_dedup=False,
        item_dedup=True,
        download_delay=2.0,
        autothrottle_start=2.0,
        autothrottle_max=60.0,
    ),

    # GitHub API spider: no HTTP-level dedup needed.
    "github_spider": BloomProfile(
        expected_urls=10_000,
        fp_rate=0.001,
        persist_dir="scrapy_researchers/.bloomstate",
        request_dedup=False,
        item_dedup=True,
        download_delay=0.0,
        autothrottle_start=0.0,
        autothrottle_max=0.0,
    ),
}

# Fallback for any spider not explicitly profiled
DEFAULT_BLOOM_PROFILE = BloomProfile(
    expected_urls=100_000,
    fp_rate=0.001,
    persist_dir="scrapy_researchers/.bloomstate",
)


def get_bloom_profile(spider_name: str) -> BloomProfile:
    """Look up the bloom profile for a spider, falling back to defaults."""
    return BLOOM_PROFILES.get(spider_name, DEFAULT_BLOOM_PROFILE)

File: .jade/crawlers/bloom-filter.ts
// .jade/crawlers/bloom-filter.ts — Bloom filter for Crawlee URL deduplication
//
// Uses the deterministic bloom filter implementation to track seen URLs.
// Integrates with Neon Postgres for cross-restart persistence via bloom_filter_state table.
//
// Usage with Crawlee:
//   const filter = new PersistentBloomFilter({ expectedItems: 10000, falsePositiveRate: 0.001 });
//   await filter.loadFromDb(pool, 'crawlee-docs', 'docs.anthropic.com');
//   // In router: if (filter.has(url)) skip; else filter.add(url);
//   await filter.saveToDb(pool, 'crawlee-docs', 'docs.anthropic.com');

import { createHash } from 'node:crypto';

// ─── Core Bloom Filter ─────────────────────────────────────────────────────

export interface BloomFilterConfig {
  /** Expected number of items to insert */
  readonly expectedItems: number;
  /** Desired false positive rate (e.g. 0.001 = 0.1%) */
  readonly falsePositiveRate: number;
}

export class BloomFilter {
  readonly bitSize: number;
  readonly hashCount: number;
  private bits: Uint8Array;
  private _itemCount: number = 0;

  constructor(config: BloomFilterConfig) {
    // Optimal bit array size: m = -(n * ln(p)) / (ln(2)^2)
    this.bitSize = Math.ceil(
      -(config.expectedItems * Math.log(config.falsePositiveRate)) / (Math.LN2 * Math.LN2)
    );
    // Optimal hash count: k = (m/n) * ln(2)
    this.hashCount = Math.ceil((this.bitSize / config.expectedItems) * Math.LN2);
    this.bits = new Uint8Array(Math.ceil(this.bitSize / 8));
  }

  /** Number of items inserted */
  get itemCount(): number {
    return this._itemCount;
  }

  /** Add a URL/string to the filter */
  add(item: string): void {
    const hashes = this._getHashes(item);
    for (const h of hashes) {
      const idx = h % this.bitSize;
      this.bits[Math.floor(idx / 8)]! |= (1 << (idx % 8));
    }
    this._itemCount++;
  }

  /** Check if a URL/string might be in the filter (may have false positives) */
  has(item: string): boolean {
    const hashes = this._getHashes(item);
    for (const h of hashes) {
      const idx = h % this.bitSize;
      if (!(this.bits[Math.floor(idx / 8)]! & (1 << (idx % 8)))) {
        return false; // Definitely not in the set
      }
    }
    return true; // Probably in the set
  }

  /** Serialize the filter to a Buffer for persistence */
  serialize(): Buffer {
    return Buffer.from(this.bits);
  }

  /** Restore the filter from a serialized Buffer */
  static deserialize(
    buf: Buffer,
    bitSize: number,
    hashCount: number,
    itemCount: number,
  ): BloomFilter {
    const filter = Object.create(BloomFilter.prototype) as BloomFilter;
    (filter as { bitSize: number }).bitSize = bitSize;
    (filter as { hashCount: number }).hashCount = hashCount;
    filter.bits = new Uint8Array(buf);
    filter._itemCount = itemCount;
    return filter;
  }

  /** Generate k hash values using double hashing: h(i) = h1 + i*h2 */
  private _getHashes(item: string): number[] {
    const h1Buf = createHash('sha256').update(item).digest();
    const h2Buf = createHash('sha256').update(item + '\x00').digest();
    const h1 = h1Buf.readUInt32BE(0);
    const h2 = h2Buf.readUInt32BE(0);

    const hashes: number[] = [];
    for (let i = 0; i < this.hashCount; i++) {
      hashes.push(Math.abs((h1 + i * h2) >>> 0));
    }
    return hashes;
  }
}

// ─── Persistent Bloom Filter (Neon Postgres backed) ────────────────────────

export interface DbPool {
  query(text: string, params?: unknown[]): Promise<{ rows: Record<string, unknown>[] }>;
}

export class PersistentBloomFilter {
  private filter: BloomFilter;
  private readonly config: BloomFilterConfig;

  constructor(config: BloomFilterConfig) {
    this.config = config;
    this.filter = new BloomFilter(config);
  }

  get itemCount(): number { return this.filter.itemCount; }
  get bitSize(): number { return this.filter.bitSize; }
  get hashCount(): number { return this.filter.hashCount; }

  add(item: string): void { this.filter.add(item); }
  has(item: string): boolean { return this.filter.has(item); }

  /** Load filter state from Neon Postgres bloom_filter_state table */
  async loadFromDb(pool: DbPool, crawlerId: string, domain: string): Promise<boolean> {
    const result = await pool.query(
      `SELECT filter_bytes, bit_array_size, hash_functions, items_inserted
       FROM bloom_filter_state
       WHERE crawler_id = $1 AND domain = $2`,
      [crawlerId, domain],
    );

    if (result.rows.length === 0) return false;

    const row = result.rows[0]!;
    this.filter = BloomFilter.deserialize(
      row['filter_bytes'] as Buffer,
      row['bit_array_size'] as number,
      row['hash_functions'] as number,
      row['items_inserted'] as number,
    );
    return true;
  }

  /** Save filter state to Neon Postgres bloom_filter_state table */
  async saveToDb(pool: DbPool, crawlerId: string, domain: string): Promise<void> {
    await pool.query(
      `INSERT INTO bloom_filter_state
         (crawler_id, domain, expected_items, false_positive_rate,
          hash_functions, bit_array_size, items_inserted, filter_bytes)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
       ON CONFLICT (crawler_id, domain) DO UPDATE SET
         items_inserted = $7,
         filter_bytes = $8,
         updated_at = NOW()`,
      [
        crawlerId, domain,
        this.config.expectedItems, this.config.falsePositiveRate,
        this.filter.hashCount, this.filter.bitSize,
        this.filter.itemCount, this.filter.serialize(),
      ],
    );
  }
}
